feat: add rdap command for domain, ip and asn lookup

BruRdgz opened the pull request:
Closes #288

BruRdgz committed:
feat: add rdap command for domain, ip and asn lookup

BruRdgz commented:
ok bro

IlluminatiFish commented:
> Lock in bro, stop trolling or I'll take more time next time

Fine by me.

Robin5605 commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,219 @@
+        if not clean_data:
+            return "No data available."
+
+        max_key_len = max(len(k) for k in clean_data)
```
```suggestion
        max_key_len = max(map(len, clean_data))
```

Robin5605 commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,219 @@
+        lines.append(f"{'-' * max_key_len}-|{'-' * 25}")
+
+        for key, value in clean_data.items():
+            lines.append(f"{key.ljust(max_key_len)} | {value}")
```
```suggestion
            lines.append(f"{key:<max_key_len} | {value}")
```

Robin5605 commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,219 @@
+                    return
+
+                data = await response.json()
+        except Exception as e:
```
this except block will never go off for RDAP errors because by default. aiohttp doesn't raise errors. you most likely want to handle this as the `else:` case for the above if-statements

Robin5605 commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,219 @@
+                result_data["Nameservers"] = ", ".join(ns_list[:3]) + ("..." if len(ns_list) > 3 else "")
+
+        elif query_type == "ip":
+            model = RDAPIP(**data)
```
```suggestion
            model = RDAPIP.model_validate(data)
```

Robin5605 replied:
Hm, I wasn't aware they did the same thing. If that's the case, then I suppose we can leave it up to personal preference - though `model_validate` seems more explicit to me.

BruRdgz commented on src/bot/constants.py:
```diff
@@ -8,7 +8,7 @@
 from typing import ClassVar
 
-from pydantic import root_validator
+from pydantic import model_validator
```
[See previous message](https://github.com/vipyrsec/bot/pull/308#discussion_r2565431756)

Robin5605 replied:
#309

BruRdgz commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,219 @@
+    """Model for Autonomous System Number RDAP responses."""
+
+    startAutnum: int | None = None
+    endAutnum: int | None = None
```
I… don’t think it would be interesting, no.
Consider [RFC 9083](https://datatracker.ietf.org/doc/rfc7483/), which this implementation is based on.
According to [RFC 9083 §5.4 — Autonomous System Number Object](https://datatracker.ietf.org/doc/html/rfc9083#section-5.4), the JSON response **must** use the keys `startAutnum` and `endAutnum` to represent the interval:
> “The Autonomous System Number object class is used to represent information regarding an Autonomous System number registration… The following members are defined:
> – `startAutnum`: the starting number
> – `endAutnum`: the ending number”

I’m following a 1:1 mapping of the RFC’s naming conventions.
If we were to make the field names consistent with our internal style, we’d need to introduce aliases (e.g., `Field(alias="startAutnum", ...)`) and possibly  serialization logic. 
I haven’t looked deeply enough to know exactly how much that would add.
I’ll defer to you on what the priority should be here.

BruRdgz commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,219 @@
+        title = f"RDAP Lookup: {query}"
+
+        if query_type == "domain":
+            model = RDAPDomain(**data)
```
[See previous message](https://github.com/vipyrsec/bot/pull/308#discussion_r2566650618)

BruRdgz committed:
Resolve reviews

Robin5605 committed:
Merge branch 'main' into bru/feat/rdap

Signed-off-by: Robin <74519799+Robin5605@users.noreply.github.com>

IlluminatiFish committed:
Lint N815

Linting N815 concerns

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

IlluminatiFish committed:
Fix N815 concerns

Fix N815 concerns

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

IlluminatiFish committed:
Lint ANN204

Add return type annotation for special method __init__

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

IlluminatiFish committed:
Lint E501

Lint E501

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

IlluminatiFish committed:
Lint E501

Lint E501

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

IlluminatiFish committed:
Lint ANN401

Lint ANN401

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

IlluminatiFish committed:
Update rdap.py

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

IlluminatiFish committed:
Ignore C901 and PLR0912 linter issues

Ignore C901 and PLR0912 linter issues, can fix them if really needed in the future.

Signed-off-by: IlluminatiFish <45714340+IlluminatiFish@users.noreply.github.com>

BruRdgz committed:
Fix lint

import-pandas-as-numpy committed:
Merge remote-tracking branch 'origin/main' into bru/feat/rdap

import-pandas-as-numpy committed:
Harden RDAP lookup handling

import-pandas-as-numpy reviewed (dismissed):
Reviewed the updated RDAP implementation against #288. ASN paths, RFC field aliases, related-response merging, nested contacts, CIDR output, error handling, and outbound URL safety are covered; live lookups and all local strict gates pass.

import-pandas-as-numpy committed:
Handle RDAP timeouts and oversized output

import-pandas-as-numpy commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,353 @@
+        except RDAPNotFoundError:
+            await ctx.send(f"❌ No results found for `{normalized_query}`.")
+            return
+        except (aiohttp.ClientError, RDAPResponseError, ValidationError):
```
Addressed in cd9a895. TimeoutError is now part of the RDAP-specific exception path, and a command-level regression test verifies upstream timeouts return the intended user-facing service error.

import-pandas-as-numpy commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,353 @@
+
+        embed = Embed(
+            title=f"RDAP Lookup: {normalized_query}",
+            description=format_table(result_data),
```
Addressed in cd9a895. Table formatting now occurs inside the RDAP-specific exception boundary, and a command-level regression test exercises a valid multi-field response that exceeds the embed table limit.

import-pandas-as-numpy committed:
Merge remote-tracking branch 'origin/main' into bru/feat/rdap

import-pandas-as-numpy committed:
Bound RDAP embed titles

import-pandas-as-numpy commented on src/bot/exts/rdap.py:
```diff
@@ -0,0 +1,354 @@
+            return
+
+        embed = Embed(
+            title=f"RDAP Lookup: {normalized_query}",
```
Addressed in the latest branch commit. RDAP embed titles are now capped at Discord’s 256-character limit with an ellipsis, and a command-level regression test exercises a valid 253-character domain.

import-pandas-as-numpy reviewed (approved):
Re-reviewed at 9b05c75 after integrating current main and addressing all Greptile feedback. All strict local hooks and 39 tests pass; every GitHub check passes; Greptile reports 5/5 with no blocking findings.
